auto-value-gson-extension/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -555,24 +556,122 @@ private static void addBuilderFieldSetting(Builder block,
+            addAllPutAll,
+            adapter,
+            jsonReader);
+      }
+      return;
+    }
+    // We're looking for a method that we can call with a parameter that is a value of the property
+    // type. The parameter type might be exactly the same, but it could also be a supertype.
+    // For example, AutoValue allows you to set a property of type ImmutableList<String> using a
+    // parameter of type List<String> (it generates a call to ImmutableList.copyOf).
+    // Similarly, it allows you to set a property of type Optional<String> with a parameter of type
+    // String (it generates a call to Optional.of). So we look for those cases.
+    // There's a further quirk, which is that the T in @AutoValue class Foo<T> is not the same as
+    // @AutoValue.Builder class Foo.Builder<T>, so if the return type of Foo.getBar() is
+    // ImmutableList<T> then that won't look the same as the parameter type of
+    // Foo.Builder.setBar(ImmutableList<T>). AutoValue has some tricky logic to juggle the types
+    // but here we just assume the types will work if their erasures do. Otherwise AutoValue would
+    // presumably reject the getter/setter combination with a helpful error message.
+    Types typeUtils = processingEnv.getTypeUtils();
+    TypeMirror propertyType = prop.element.getReturnType();
+    for (ExecutableElement setter : setters) {
+      if (typeUtils.isAssignable(
+              typeUtils.erasure(propertyType), typeUtils.erasure(setter.getParameters().get(0).asType()))) {
+        block.addStatement(
+            "$N.$N($N.read($N))", builder, setter.getSimpleName(), adapter, jsonReader);
+        return;
+      }
+    }
+    Optional<TypeMirror> optionalArgType = optionalArgType(propertyType, typeUtils);
+    if (optionalArgType.isPresent()) {
+      for (ExecutableElement setter : setters) {
+        if (typeUtils.isAssignable(
+            typeUtils.erasure(optionalArgType.get()), typeUtils.erasure(setter.getParameters().get(0).asType()))) {
+          // The AutoValue property is `Optional<String> foo()` but we only have a setter
+          // `setFoo(String)`.
+          // We'll generate `adapter.read(jsonReader).ifPresent(x$ -> builder.setFoo(x$))`.
+          block.addStatement(
+              "$N.read($N).ifPresent(x$$ -> $N.$N(x$$))",
+              adapter,
+              jsonReader,
+              builder,
+              setter.getSimpleName());
+          return;
+        }
+      }
+    }
+    // This shouldn't happen: AutoValue should have rejected the builder if it didn't have a way to
+    // set this property. But it may be that AutoValue has output an error message and still let
+    // this extension run.
+    String errorMsg = "Setter not found for " + prop.element;
+    processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, errorMsg, prop.element);
+  }
+
+  private static Optional<TypeMirror> optionalArgType(TypeMirror typeMirror, Types typeUtils) {
```

Comment:
It's worth mentioning we don't technically support `Optional` properties in auto-value-gson (#161, #123, https://github.com/google/auto/issues/359). If it's working, it's by coincidence and not intentional :). Does that obviate the need for much of the below or do you think it's worth keeping this around as a toe-hold for better supporting it in the future?

Reply:
The background to this PR is that I am trying to update Google's giant internal [monorepo](https://cacm.acm.org/magazines/2016/7/204032-why-google-stores-billions-of-lines-of-code-in-a-single-repository/fulltext) to a more recent version of auto-value-gson. We're currently on v0.8.0 so a lot has changed since then. I've basically imported the current code to my own workspace, and I'm working on fixing the things that that change breaks before I can submit it to the monorepo. The version we have predates the use of builders during deserialization, so some things that happened to work then don't work now. This PR is an attempt to address some of those things, in a way that is useful in general and not just to Google.

I haven't really studied the question of `Optional` specifically in detail, but this change does help with some of the cases I just mentioned. Yes, google/auto#359 would make some of the code here a lot simpler if it were addressed, but there's some API review and release choreography that would have to happen before we could make that simplification.

In answer to your other question, I'm importing the code from the current `main` branch. I don't need a release for this to be useful to me.

Reply:
Great, I'll merge this then. @cushon lookee into this awhile back iirc and ran into issues with our past removal of mutable adapters + default value setting, fyi

Reply:
Thanks! Yes, I'm running into those same issues too, but I think I'll be able to fix them by modifying the application code in question rather than by tweaking auto-value-gson.